# CS 120 - Introduction to Programming - Fall 2021

## Class and Instructor Details

CS 120-004 - Introduction to Programming (3 credits)

### Instructor

Virgil U Pierce <br/>
Email: virgil.pierce@unco.edu <br/>
Cell: 956-249-0566

Office: Ross Hall 2239

### Class Meetings

Monday, Wednesday, and Friday: 10:10am to 11:00am <br/>
Ross Hall 2261

### COVID-19 Adjustments

I suggest wearing a mask. Incidence of breakthrough infections are non-zero, and the consequence of an infection appear to be severe and potentially long term even in young people (and I for one am not so young anymore). 

My wife has a medical condition making vaccines generally less effective for her, and I therefore respectfully ask that you not attend class if you have a fever or cough. I will make arrangements for people to join us via Zoom if needed.

### Student Hours

Online and To be Determined.

Appointments can be made via:  [https://virgilpierce.youcanbook.me](https://virgilpierce.youcanbook.me)

## Course Description

From the catalogue: Become familiar with designing and writing programs in a high-level language. Programs will involve some technical applications in science and mathematics. 

Topics will include:

- Built-in variable types:  Integers, Floating Point Numbers, Lists, Tuples, and Dictionaries.
- Loops and Recursion
- Functions
- Conditions and Flow Control
- User Defined Types
- Using common modules for math and science computations

## Software

All of the software (and the book) for class is available for free:

### Python

The course uses Python as the primary programming langauge. Python is an excellent first language, and also a good environment for exploring algorithms before taking them to other languages for implementation.

Python's advantages include:  
- A syntax that encourages writing easy to read programs
- human readable code:  almost always a statement does exactly what it says it does if you read it in English
- dynamic typing: typing variables is handled automatically removing one of the major barriers in learning a first language
- Interactive Dynamic Environment: Python can be run in an IDE meaning you can execute a statement and then see what it did

Python is also now the most commonly used language in math and science for professional computations. UNC is in the process of phasing out the use of commercial software for mathematics computations to use Python instead. 

The disadvantages of Python are:
- It is dynamically typed:  Your code will not give you a warning if you assume something is an integer and it is not.
- It is not usually compiled:  Compiling is a process other languages use to process the human written code into instructions for the computer. Compiled languages are much much faster.

### Jupyter

There are a variety of IDE softwares you could use which provide the primary interface we will use for interacting with Python. I am particularly fond of Jupyter for reasons we will discuss in class. You have the following options for accessing Jupyter:

1. You can install a local version of Jupyter on your personal computer. I recommend downloading Anaconda: https://www.anaconda.com/products/individual which is a package manager for Python / Jupyter. 

2. You can access something like Jupyter with Google Colab. This does require a google account; but you will not need to install anything, and it will work on Chromebooks, tablets, and phones.

3. Jupyter is also isntalled on the university Apporto Server which you can access here: https://unco.apporto.com/  You will not need to install anything, and it will work on Chromebooks, tablets, and phones.

4. If you are interested, we have a set of computers the university is letting us install Linux on, and once we get them configured for Linux we will be able to have Jupyter/Python running there.


## Textbook

We will be using the *Think Python* book by Allen Downey. It is available for free here: https://greenteapress.com/wp/think-python-2e/

You can purchase a print copy if you prefer that. 

### Course Notes

The notes we use in course, include some code examples are provided on Github: https://github.com/virgilpierce/CS_120

If you are in the Applied Math or Software Engineering majors it would be worth your time to start learning how to use Github, though for our course I am not going to require that you use it.

### Other References

There are a lot of resources for Python available. If you would like more detail, I recommend the O'Reily series of books. However I also find that books rarely answer all of my questions - learning how to google for information about how to do things or what is not working is an important skill to develop as a scientist and a programmer specifically.

There is some inline help if you are using the IDE or Jupyter:

In [1]:
# Help page describing the function or type

int?

Init signature: int(self, /, *args, **kwargs)
Docstring:     
int([x]) -> integer
int(x, base=10) -> integer

Convert a number or string to an integer, or return 0 if no arguments
are given.  If x is a number, return x.__int__().  For floating point
numbers, this truncates towards zero.

If x is not a number or if base is given, then x must be a string,
bytes, or bytearray instance representing an integer literal in the
given base.  The literal can be preceded by '+' or '-' and be surrounded
by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
Base 0 means to interpret the base from the string as an integer literal.
>>> int('0b100', base=0)
4
Type:           type
Subclasses:     bool, IntEnum, IntFlag, _NamedIntConstant


In [2]:
# A list of the non system attributes of a function, type, or object

[x for x in dir(int) if not '__' in x]

['as_integer_ratio',
 'bit_length',
 'conjugate',
 'denominator',
 'from_bytes',
 'imag',
 'numerator',
 'real',
 'to_bytes']

## Learning Objectives / Outcomes for the Course

As a result of this course, students will:

1. use computer programming to solve computational problems in mathematics and science;
2. develop algorithms and then implement them
3. evaluate code written by someone else and explain what the result should be, or what correction is needed.
4. write functions, loops, and control structures.

A typical week of class will involve posing a question for which Python is a good solution, exploring some of the tools we need in order to solve it, and then learning how to use those tools as implemented for Python.

-----

A caveate about the course and this topic:  I have learned 8 computer languages fluently (and 0 human languages - I know just enough Spanish to get into a lot of trouble in Mexico). Each one you learn will be easier than the ones before with some exceptions. I have never learned them from a course - the best way to learn the language is to find a problem you want to solve and then write a program to solve it. Do enough of this and you will know the language.

**So why this course?**  I am going to pose to you a series of problems that will help us formulate structure we need from Python to solve it; we will then learn about that structure, and then solve the problem. In the process we will be learning Python, but more generally we will be learning about how algorithms a computer can use are structured and how to build an algorithm for a problem. 

Future courses in CS as UNC you can take will cover more powerful tools, particularly ones that exploit the way computers work, and the analysis of algorithms to determine which ones are best and estimate how much power is needed.

This is a fascinating subject. If you are interested in languages beyond Python and Java that we teach, I would like to be forming a club for students that are learning newer languages that are less used.

In [3]:
import CS_120_Schedule
CS_120_Schedule.schedule

Day   Week_Day                                            Title  \
0   2021-08-23     Monday                 Introduction to Python / Jupyter   
1   2021-08-25  Wednesday                          Variables and Functions   
2   2021-08-27     Friday                     Lab Day: Euclidean Algorithm   
3   2021-08-30     Monday                       Conditionals and Recursion   
4   2021-09-01  Wednesday                       Conditionals and Recursion   
5   2021-09-03     Friday                    Lab Day: Find N! and (2N-1)!!   
6   2021-09-06     Monday                    No Class - University Holiday   
7   2021-09-08  Wednesday                       Local and Global Variables   
8   2021-09-10     Friday    Lab Day: Exploring Local and Global Variables   
9   2021-09-13     Monday                   Limitations of Floats and Ints   
10  2021-09-15  Wednesday                              Dealing with Errors   
11  2021-09-17     Friday              Lab Day:  Approximate a square root   
12  2021-09-20     Monday                                More Using Module   
13  2021-09-22  Wednesday                             Calculus with Python   
14  2021-09-24     Friday           Lab Day: How fast does factorial grow?   
15  2021-09-27     Monday                                       Iterations   
16  2021-09-29  Wednesday                                       Iterations   
17  2021-10-01     Friday                              Lab Day: Iterations   
18  2021-10-04     Monday                                       Iterations   
19  2021-10-06  Wednesday                                          Strings   
20  2021-10-08     Friday                                 Lab Day: Strings   
21  2021-10-11     Monday                                          Strings   
22  2021-10-13  Wednesday                                          Strings   
23  2021-10-15     Friday                        Lab Day: Analyzing Hamlet   
24  2021-10-18     Monday                                 Lists and Tuples   
25  2021-10-20  Wednesday                                 Lists and Tuples   
26  2021-10-22     Friday  Lab Day: How many primes are there less than N?   
27  2021-10-25     Monday                                 Lists and Tuples   
28  2021-10-27  Wednesday                                 Lists and Tuples   
29  2021-10-29     Friday    Lab Day: Find the derrivative of a polynomial   
30  2021-11-01     Monday                            Dictionaries and Sets   
31  2021-11-03  Wednesday                            Dictionaries and Sets   
32  2021-11-05     Friday                 Lab Day: Algebra of Permutations   
33  2021-11-08     Monday                          Putting it All Together   
34  2021-11-10  Wednesday                          Putting it All Together   
35  2021-11-12     Friday                          Putting it All Together   
36  2021-11-15     Monday                         Data Structure Selection   
37  2021-11-17  Wednesday                         Data Structure Selection   
38  2021-11-19     Friday   Lab Day: Complex Numbers, Quaternionic Numbers   
39  2021-11-22     Monday                                File Manipulation   
40  2021-11-24  Wednesday                    No Class - University Holiday   
41  2021-11-26     Friday                    No Class - University Holiday   
42  2021-11-29     Monday                             Lab Day: Using Files   
43  2021-12-01  Wednesday                               Things left undone   
44  2021-12-03     Friday                               Things left undone   
45  2021-12-06     Monday                               Final Exam/Project   

                                 Description                  Notes  
0                           Variables, Numpy    Euclidean Algorithm  
1   Types of variables and writing functions                         
2           Implementing Euclidean Algorithm                         
3             Fixing our Euclidean Algorithm                    

## Course Outline

Python is quickly replacing a lot of other types of software I use in my daily work. In particular it has some big advantages as a replacement for spreadsheets. I am using it here to make the schedule for our class:

## Connections with Industrial Mathematics

This course is directly connected with the growing mathematics field of Data Science. There are many opportunities for people with an expertise in this field. However the field is very accessible so you have competition.

Some thoughts on how to rise to the top:
1. Have some projects you have done that you can talk about,
2. Be able to talk about working with a team i.e. how do you help your team be effective, how did you get your team over a challenge, how do you lead?,
3. Be able to talk about challenges in the field (we will do this),
4. Have a breadth of experience in science or business generally, and mathematics, statistics, and computer science specifically (we will do this), and
5. The more you have done the better - I suggest devoting some of your play time to working with data (we will explore Kaggle in this class, and my suggestion is that it become one of your hobbies).

I have some literature on building a career in Data Science available if you like.

If you are interested in this direction consider:  Any of the other CS courses we offer, the MATH 351 and STAT 409 and 411 courses.

## Assessments

As with any math or science class, doing the work is essential to processing and learning the material. Furthermore the concepts covered in this class are not meant to be purely abstract things, but instead are actual tools deployable to solve computational problems or to issue instructions to a computer to display or record information.  You must do the activities from class and homework in order to make progress.

### Homework

There will be weekly homework assignments asking you to either write code or evaluate someone elses code.  These are indended to be done with teammates and I highliy recommend having a group from class that you want to work with. We have space in the Ross 22XX wing if your group needs a place to work and needs computers to work on.

### Checking your understand

We will do a number of activites during class as teams to help you check your understanding or develop new understandings. 

### Class Presentations and Flipgrid

We will use flipgrid as part of class to have you share your reasoning and methods with others in the class.

### Projects

There will be a few projects you will do as part of class that ask you to put together multiple concepts form class to solve a computational problem (or more correctly to have a computer solve the problem). Part of your project will be a video explaining your approach and how it is working.

### Final

We will have a final exam with some questions you have been prepped on ahead of time. They will not involve writing new code so much as fixing someone elses code or identifying what the bug is. 

### Grading and Feedback

You will have a chance to submit corrections for the homework and projects. Your final grade will be based on a weighting of:  Homework 30%; Projects 30%, Final 10%, Checking your Understanding 15%, and presentations 15%. However I am most interested in seeing that you have made progress during the course. 



## UNCO Policy Statements

### Disability Resources

It is the policy and practice of the University of Northern Colorado to create inclusive learning environments. If there are aspects of the instruction or design of this course that present barriers to your inclusion or to an accurate assessment of your achievement such as time-limited exams, inaccessible web content, or use of videos without caption, please communicate this with your professor and contact Disability Support Services (DSS) at (970) 351-2289, Michener Library L-80 to request accommodations. Students can learn more about the accommodation process at  [https://www.unco.edu/disability-resource-center/accommodations/](https://www.unco.edu/disability-resource-center/accommodations/)

### Food Insecurity and Basic Needs

Research shows that college students experience food insecurity at higher rates than the American household rate, and that food insecurity can negatively impact academic performance and persistence. In recognition of this problem, UNC offers assistance to students facing food insecurity through an on- campus food pantry. The Bear Pantry is located in University Center 2166A, and is open for regular hours throughout the semester.  

Please visit [www.unco.edu/bear-pantry](www.unco.edu/bear-pantry) for more information.

Any student who faces challenges securing their food or housing and believes this may affect their performance in the course is also urged to contact Student Outreach and Support (SOS) for assistance.  SOS can assist students during difficult circumstances which may include medical, mental health, personal or family crisis, illness or injury. SOS can be reached at sos@unco.edu or via phone at 970-351-2796.

### Honor Code

All members of the University of Northern Colorado community are entrusted with the responsibility to uphold and promote five fundamental values: Honesty, Trust, Respect, Fairness, and Responsibility. These core elements foster an atmosphere, inside and outside of the classroom, which serves as a foundation and guides the UNC community's academic, professional, and personal growth. Endorsement of these core elements by students, faculty, staff, administration, and trustees strengthens the integrity and value of our academic climate.

Passing off others work as your own is very unethical and when it comes to computer code is also a security problem for your machine and your data. All assignments will clearly identify to what extent collaboration is allowed, and in the case of homework it is encouraged. If in doubt, please do ask, and always provide attribution if you are using what someone else has done. Interestingly, we will be using a few of the standard modules for Python which are by definition code others have written, but we will be explicit in what we are doing. 

### UNC's Policies

UNC's policies and recommendations for academic misconduct will be followed. For additional information, please see the Student Code of Conduct at the Dean of Student’s website [http://www.unco.edu/dos/Conduct/codeofconduct.html](http://www.unco.edu/dos/Conduct/codeofconduct.html). In the case of academic appeals, university procedures will be followed. For information on academic appeals, see [https://www.unco.edu/trustees/pdf/board-policy-manual.pdf](https://www.unco.edu/trustees/pdf/board-policy-manual.pdf).